shaders/src/fog.fs

```diff
@@ -2,43 +2,49 @@
-
-        // clamp to camera far plane so the skybox participates to the fog
-        float cameraToWorldDistanceClamped = min(frameUniforms.cameraFar, cameraToWorldDistance);
-
-        // fog density with respect to camera's altitude
-        float density = frameUniforms.fogDensity * exp2(-frameUniforms.fogHeightFalloff * (wsCameraHeight - frameUniforms.fogHeight));
-
-        // falloff due to height of the shaded fragment w.r.t. the camera
-        float h = dot(view, up);
-        h = abs(h) > 0.01 ? h : 0.01;
-        float falloff = max(frameUniforms.fogHeightFalloff * h, -127.0);
-
-        // fog density at shaded fragment w.r.t. camera's altitude
-        float densityFalloff = density * (1.0 - exp2(-falloff)) / falloff;
-
-        // take fog start distance into account
-        float lineIntegral = densityFalloff * max(cameraToWorldDistanceClamped - frameUniforms.fogStart, 0.0);
-        float fogFactor = max(clamp(exp2(-lineIntegral), 0.0, 1.0), 1.0 - frameUniforms.fogMaxOpacity);
-
-        vec3 foggedColor = frameUniforms.fogColor * (1.0 - fogFactor);
-        if (frameUniforms.fogInscatteringSize > 0.0) {
-            vec3 lightDirectionNorm = -normalize(frameUniforms.lightDirection);
-            vec3 cameraToWorldUnit = view * (1.0 / cameraToWorldDistance);
-            float lightInscatteringFactor = max(dot(cameraToWorldUnit, lightDirectionNorm), 0.0);
-            vec3 lightInscattering = frameUniforms.lightColorIntensity.rgb * frameUniforms.lightColorIntensity.w * pow(lightInscatteringFactor, frameUniforms.fogInscatteringSize);
-            float lightInscatteringIntegral = densityFalloff * max(cameraToWorldDistanceClamped - frameUniforms.fogInscatteringStart, 0.0);
-            float inscatteringFactor = clamp(exp2(-lightInscatteringIntegral), 0.0, 1.0);
-            foggedColor += lightInscattering * (1.0 - inscatteringFactor);
+        float A = frameUniforms.fogDensity;
+        float B = frameUniforms.fogHeightFalloff;
+
+        float d = length(view);
+        float h = abs(view.y) < FLT_EPS ? FLT_EPS : view.y;   // avoid divide-by-zero
+
+        float fogIntegralFunctionOfDistance = A * (1.0 - exp2(-B * h)) / h;
+        float fogIntegral = fogIntegralFunctionOfDistance * max(d - frameUniforms.fogStart, 0.0);
+
+        // remap fogIntegral to an opacity between 0 and 1. saturate(fogIntegral) is too harsh.
+        float fogOpacity = max(1.0 - exp2(-fogIntegral), 0.0);
+        //float fogOpacity = clamp(fogIntegral, 0.0, 1.0);
+
+        // don't go above requested max opacity
+        fogOpacity = min(fogOpacity, frameUniforms.fogMaxOpacity);
+
+        // compute fog color
+        vec3 fogColor = frameUniforms.fogColor;
+
+        if (frameUniforms.fogColorFromIbl > 0.0) {
+            // get fog color from envmap
+            float lod = frameUniforms.iblRoughnessOneLevel;
+            fogColor = textureLod(light_iblSpecular, view, lod).rgb * frameUniforms.iblLuminance;
```

Comment:
`*=` instead of `=`? That way you can tint IBL fog by using a color != 1,1,1.

Reply:
yeah I was wondering about that.